CalibTracker/SiStripChannelGain/src/SiStripGainsPCLHarvester.cc

```diff
@@ -0,0 +1,440 @@
+	    APV->Gain          = -1;
+	    APV->PreviousGain  = 1;
+	    APV->PreviousGainTick = 1;
+	    APV->x             = DetUnit->position().basicVector().x();
+	    APV->y             = DetUnit->position().basicVector().y();
+	    APV->z             = DetUnit->position().basicVector().z();
+	    APV->Eta           = DetUnit->position().basicVector().eta();
+	    APV->Phi           = DetUnit->position().basicVector().phi();
+	    APV->R             = DetUnit->position().basicVector().transverse();
+	    APV->Thickness     = DetUnit->surface().bounds().thickness();
+	    APV->isMasked      = false; //SiPixelQuality_->IsModuleBad(Detid.rawId());
+	    APV->NEntries      = 0;
+	    
+	    APVsCollOrdered.push_back(APV);
+	    APVsColl[(APV->DetId<<4) | APV->APVId] = APV;
+	    Index++;
+	    NPixelDets++;
+
+	  } // loop on ROC cols
+	} // loop on ROC rows
+      } // if Pixel
+    } // loop on Dets  
+  }  //if (!bareTkGeomPtr_) ... 
+  bareTkGeomPtr_ = newBareTkGeomPtr;
+}
+
+//********************************************************************************//
+bool SiStripGainsPCLHarvester::produceTagFilter(const MonitorElement* Charge_Vs_Index){
+  
+  // The goal of this function is to check wether or not there is enough statistics
+  // to produce a meaningful tag for the DB
+
+  if( Charge_Vs_Index==0 ) {
+    edm::LogError("SiStripGainsPCLHarvester") << "produceTagFilter -> Return false: could not retrieve the "
+					      << m_calibrationMode <<" statistics.\n"
+					      << "Please check if input contains " 
+					      << m_calibrationMode << " data." << std::endl;
+    return false; 
+  }
+  
+  
+  float integral = (Charge_Vs_Index)->getTH2S()->Integral();
+  if( (Charge_Vs_Index)->getTH2S()->Integral(0,NStripAPVs+1, 0, 99999 ) < tagCondition_NClusters) {
+    edm::LogWarning("SiStripGainsPCLHarvester") 
+      << "calibrationMode  -> " << m_calibrationMode << "\n"
+      << "produceTagFilter -> Return false: Statistics is too low : " << integral << std::endl;
+    return false;
+  }
+  if((1.0 * GOOD) / (GOOD+BAD) < tagCondition_GoodFrac) {
+    edm::LogWarning("SiStripGainsPCLHarvester")
+      << "calibrationMode  -> " << m_calibrationMode << "\n"
+      << "produceTagFilter ->  Return false: ratio of GOOD/TOTAL is too low: " << (1.0 * GOOD) / (GOOD+BAD) << std::endl;
+    return false;
+  }
+  return true; 
+}
+
+//********************************************************************************//
+std::unique_ptr<SiStripApvGain>
+SiStripGainsPCLHarvester::getNewObject(const MonitorElement* Charge_Vs_Index) 
+{
+  std::unique_ptr<SiStripApvGain> obj = std::unique_ptr<SiStripApvGain>(new SiStripApvGain());
+  
+  if(!produceTagFilter(Charge_Vs_Index)){
+    edm::LogWarning("SiStripGainsPCLHarvester")<< "getNewObject -> will not produce a paylaod because produceTagFilter returned false " << std::endl;       
+    return obj;
+  } else {
+    doStoreOnDB=true;
+  }
+  
+  std::vector<float>* theSiStripVector = NULL;
```

Comment:
it seems that all of these vectors are leaked. why do they need to be pointers? (and in general the complex logic below could be simplified by checking for empty vectors and clearing the vector when a new detid is found